Fix #17523: Github Team Discussion deprecation

SD-13 opened the pull request:
Resurrecting @462702985's  [PR](https://github.com/oppia/stale-review-request-notifier/pull/1).
This PR will fix https://github.com/oppia/oppia/issues/17523 by setting up this repo which will be used to notify reviewers of pending reviews.

### Proof of Changes
Before pushing the major changes, I am testing the functionality in my repos. The [pending-review-notifier](https://github.com/SD-13/pending-review-notifier) contains the code and in [test-pending-review-notifier](https://github.com/SD-13/test-pending-review-notifier) repo, I used it as a GitHub action. To see the runs and the comments please check the Actions and the Discussions of [test-pending-review-notifier](https://github.com/SD-13/test-pending-review-notifier) repo.

SD-13 committed:
initial changes

SD-13 commented:
@seanlip @gp201 I am thinking about opening a separate PR to add workflows for lint checks, mypy, and unit tests first, I think it will help me and you guys to review this PR with the tests in GitHub action. What do you think?

SD-13 committed:
few more changes

SD-13 committed:
update readme

SD-13 commented:
@seanlip @gp201 Please look into this PR when you are free. Thanks!

SD-13 commented:
To test this functionality, I used two repos -
1. [pending-review-notifier](https://github.com/SD-13/pending-review-notifier) -  It worked as the GitHub action. Almost same as this PR's changes with some small tweaks( e.g changing the discussion comment with `test comment` to not bother reviewers, hardcoded the repo name in between the scripts to fetch data from `oppia/oppia` and posting comments in the below repo.
2. [test-pending-review-notifier](https://github.com/SD-13/test-pending-review-notifier) -  I used it as the repo to run the workflow.

Then if you look into the test repo, you can see [this workflow run](https://github.com/SD-13/test-pending-review-notifier/actions/runs/5342915663/jobs/9685313310) posted some comments in [this discussion](https://github.com/SD-13/test-pending-review-notifier/discussions/1). Please let me know if you have any questions in this regard. Thanks,

SD-13 commented:
Hi @DubeySandeep , Can you please review this PR. Thanks!

DubeySandeep commented:
@SD-13 Thanks for the PR, I won't be able to review this PR today, I'll take a pass tomorrow!

seanlip commented on .github/PENDING_REVIEW_NOTIFICATION_TEMPLATE.md:
```diff
@@ -0,0 +1,8 @@
+Hi {{ username }},
```
@SD-13 Could we use the updated message I have in this PR? https://github.com/oppia/oppia/pull/18553/files#diff-56bf299501545e9fd22a67dbb874edd4764f26fe0040dac349d06c95b5a9dc7eR1

Also just to check, is that file in the oppia/oppia repo still the source of truth, or is it this one? If one of them isn't used we should get rid of it. /cc @DubeySandeep

SD-13 replied:
Right now it is using the one in this repo, so we can get rid of the template in oppia/oppia.

SD-13 replied:
@seanlip I have used your message but with a small change. Also, as per the current work, this repo template will be used, so we can get rid of the one in `oppia/oppia` repo.

seanlip commented on .github/actions/send-webhook-notification/action.yml:
```diff
@@ -0,0 +1,24 @@
+runs:
+  using: "composite"
+  steps:
+    - name: Store build URL in an environment variable
```
I'm a bit confused here because I don't know what the build URL refers to or what it has to do with this PR.

SD-13 replied:
I think I added it by mistake, removed it.

seanlip commented on README.md:
```diff
@@ -0,0 +1,83 @@
+      - Replace `<< category_name >>` and `<< discussion_title >>` with the respective category name and discussion title.
+      - Replace `<< TURNAROUND_HOURS >>` with the expected PR review time.
+      - Don't use space in `category_name` or `discussion_title`; otherwise, they will be considered different arguments.
+      - The [POSIX cron syntax](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/crontab.html#tag_20_25_07) needs to quoted as * is a special character in YAML.
```
"...needs to be quoted ..."

SD-13 replied:
done

seanlip commented on README.md:
```diff
@@ -0,0 +1,83 @@
+      - Don't use space in `category_name` or `discussion_title`; otherwise, they will be considered different arguments.
+      - The [POSIX cron syntax](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/crontab.html#tag_20_25_07) needs to quoted as * is a special character in YAML.
+
+4. Add PENDING_REVIEW_NOTIFICATION_TEMPLATE.yml file in `.github/` dir.
```
Isn't that file already there? If so then why do the usage instructions say to "add" it again?

SD-13 replied:
removed this from README

seanlip commented on action.yml:
```diff
@@ -0,0 +1,30 @@
+name: Send pending review notifications to reviewer on github-discussion
+description: 'Send pending review notifications on team-discussion'
```
Is team-discussion here the same as github-discussion above? If so, let's standardize how we refer to it.

SD-13 replied:
changed it to github-discussion

seanlip commented on src/github_domain.py:
```diff
@@ -0,0 +1,111 @@
+        self,
+        name: str,
+        timestamp: datetime.datetime = DEFAULT_TIMESTAMP
+    ) -> None:
```
I think a docstring would be helpful here -- explain whether the name is the real name or github username; explain what the timestamp is. (Perhaps renaming would help to clarify.) This is because the natural concept of "assignee" does not include a timestamp, so some explanation is needed.

SD-13 replied:
updated. PTAL

seanlip replied:
But which event is the timestamp referring to?

Basically, you are saying that this class (Assignee) has two attributes, username and timestamp. In other words, the username and the timestamp are **properties** of the assignee. But if you think about it in a real-world context, it doesn't make sense for a general event timestamp to be a property of the assignee, right? That's why I don't understand the modelling here, and if you use vague docstrings like "the timestamp when an event happened", it doesn't really add much in terms of meaning.

SD-13 replied:
Updated the argument name to `assigned_on_timestamp`. This argument takes the time when the particular assignee was assigned in a PR. So, from the context of an assignee, these two attributes are making a bit of sense to me because one attribute is the username of the assignee, and the other one is the time when the assignee was assigned to a PR. I agree that it was not explained well enough but I tried to explain it this time. Please have a look and let me know if you still think that it needs changes. Thanks!

seanlip replied:
This is much better now. Thanks!

seanlip commented on src/github_domain.py:
```diff
@@ -0,0 +1,111 @@
+        self.timestamp = timestamp
+
+    def get_readable_waiting_time(self) -> str:
+        """"Returns readable waiting time on review."""
```
This should start with 3 quotes instead of 4.

What is "readable waiting time", can you explain in more understandable terms?

SD-13 replied:
updated. PTAL and let me know if that works

seanlip commented on src/github_domain.py:
```diff
@@ -0,0 +1,111 @@
+        self,
+        url: str,
+        number: int,
+        author: str,
```
Perhaps author_username is better -- an author and an assignee both sound like "people" so one would expect them to have similar types, but they don't.

SD-13 replied:
updated

seanlip commented on src/github_domain.py:
```diff
@@ -0,0 +1,111 @@
+            len(self.assignees) == 1 and self.assignees[0].name == self.author))
+
+    def get_assignee(self, user: str) -> Optional[Assignee]:
+        """Returns the assignee object for the given user if exist."""
```
...if it exists.

SD-13 replied:
done

seanlip commented on src/github_domain.py:
```diff
@@ -0,0 +1,111 @@
+        self.assignees = assignees
+
+    def is_reviewer_assigned(self) -> bool:
+        """Checks whether a reviewer assigned to the PR."""
```
...is assigned to the PR.

SD-13 replied:
done

seanlip commented on src/github_domain.py:
```diff
@@ -0,0 +1,111 @@
+        return f'PR #{self.number} by {self.author}'
+
+    # Here we use type Any because the response we get from the api call
+    # is hard to annotate in a typedDict, Hence used type Any.
```
...in a typedDict. 

(Drop the rest of it, since that part is redundant.)

SD-13 replied:
done

seanlip commented on src/github_services.py:
```diff
@@ -0,0 +1,259 @@
+    _TOKEN = token
+
+
+# Here we use type Any because.
```
This comment doesn't make sense, you might need more detail.

SD-13 replied:
updated

seanlip commented on src/github_services.py:
```diff
@@ -0,0 +1,259 @@
+    the token.
+    """
+
+    # Here we use type Any because.
```
Same problem.

SD-13 replied:
updated

seanlip commented on src/github_services.py:
```diff
@@ -0,0 +1,259 @@
+
+
+# Here we use type Any because the response we get from the api call is hard
+# to annotate in a typedDict, Hence used type Any.
```
Ditto, drop the redundant part.

SD-13 replied:
done

seanlip commented on src/github_services.py:
```diff
@@ -0,0 +1,259 @@
+    if discussion_id is None:
+        raise Exception(f'{discussion_category} category is missing in GitHub Discussion.')
+
+    query = """
```
Maybe you could add a comment explaining what's going on here a bit, with a reference to the language it's written in?

SD-13 replied:
added, please let me know if that works

seanlip commented on tests/github_services_test.py:
```diff
@@ -0,0 +1,17 @@
+
+"""Unit test for the github_services.py file."""
+
+from __future__ import annotations
```
The tests here are missing?

SD-13 replied:
Added some tests. but I am looking to add a few more cases

seanlip commented on tests/index_test.py:
```diff
@@ -0,0 +1,122 @@
+from src import github_services
+
+
+class ModuleIntegrationTest(unittest.TestCase):
```
Where are the actual tests?

SD-13 replied:
added.

seanlip reviewed (commented):
@SD-13 Thanks again for the PR. I took a pass, mostly for understandability (as mentioned, I think @DubeySandeep will be able to do a better review). But hopefully the comments help!

SD-13 committed:
some suggested changes

SD-13 committed:
update comment content

SD-13 committed:
update readme and add check for args

SD-13 committed:
mock and test post

SD-13 committed:
mock post request with errors

SD-13 committed:
to get help

SD-13 committed:
unnecessary code

SD-13 committed:
fix integration test

SD-13 committed:
add tests in index and services

SD-13 commented:
@seanlip sorry for taking this long. I have addressed your comments. Please take another pass. Thanks

seanlip commented on .github/PENDING_REVIEW_NOTIFICATION_TEMPLATE.md:
```diff
@@ -0,0 +1,8 @@
+It looks like you're assigned to this PR, but haven't taken any action for at least 2 days:
+{{ pr_list }}
+
+Please review the pending PRs as soon as possible, and reply to this thread once all the PRs are reviewed.
```
Suggested change: Please review and deassign yourself from...

(This is because sometimes people forget to deassign themselves)

SD-13 replied:
done

seanlip commented on action.yml:
```diff
@@ -0,0 +1,30 @@
+name: Send pending review notifications to reviewer on github-discussion
```
discussion --> discussions (perhaps better: GitHub Discussions)

SD-13 replied:
done

seanlip commented on src/github_domain.py:
```diff
@@ -0,0 +1,138 @@
+
+    def set_timestamp(self, timestamp: datetime.datetime) -> None:
+        """Sets timestamp to assignee.
+        Args:
```
Add a newline above this one.

SD-13 replied:
done, `outdated` is not showing here because the method name is also changed.

seanlip commented on src/github_domain.py:
```diff
@@ -0,0 +1,138 @@
+
+    def get_readable_waiting_time(self) -> str:
+        """Returns readable waiting time on review.
+        waiting time = ( current time - when reviewer was assigned )
```
Don't put code in docstrings. Instead, you can say: "The waiting time is the difference between the current time and the time when the reviewer was assigned to the PR."

SD-13 replied:
updated

seanlip commented on src/github_domain.py:
```diff
@@ -0,0 +1,138 @@
+
+    def is_reviewer_assigned(self) -> bool:
+        """Checks whether a reviewer is assigned to the PR.
+        Returns:
```
Ditto, add one newline above. Please fix all other occurrences in this PR.

SD-13 replied:
done. This time I ran the lint checks from `oppia/oppia` repo on the code in this PR.

seanlip commented on src/github_services.py:
```diff
@@ -0,0 +1,280 @@
+) -> None:
+    """Comment in the existing GitHub discussion."""
+
+    # Here this query is written in GraphQL and is being used to fetch data about the
```
Here this --> The following

SD-13 replied:
done

seanlip commented on src/index.py:
```diff
@@ -0,0 +1,147 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""The index file for sending pending-review notifications."""
```
Why is this called index.py -- in what way is it an "index"? Why not main.py or send_pending_review_notifications.py or similar?

SD-13 replied:
I have changed the file name to main for now but I am also open for suggestions.

seanlip commented on tests/github_domain_test.py:
```diff
@@ -0,0 +1,85 @@
+class AssigneeDomainUnitTest(unittest.TestCase):
+    """Assignee class test."""
+    def test_constructor_creates_object(self):
+        obj = Assignee('username', timestamp=datetime(1, 1, 1))
```
In general, try to use more descriptive variable names -- in this case, you could use `assignee`, for example. This helps readers understand better what model the test is trying to simulate. Ditto elsewhere.

SD-13 replied:
I have renamed the variables there and in other places as well. Please let me know if you think any variable name still need to be renamed.

seanlip commented on tests/github_services_test.py:
```diff
@@ -0,0 +1,235 @@
+                'login': 'reviewerName2',
+            }]
+        }]
+        def get_past_time(hours: int=0) -> str:
```
Pull this into a separate top-level private _get_past_time() method within the class, and call it using _get_past_time(...). Try not to define methods within other methods.

SD-13 replied:
done

seanlip commented on tests/index_test.py:
```diff
@@ -0,0 +1,218 @@
+
+"""Unit test for the index.py file."""
+
+import unittest
```
I think you can sort these better; I'm not sure why pylint isn't catching it?

SD-13 replied:
updated this time

seanlip commented on tests/index_test.py:
```diff
@@ -0,0 +1,218 @@
+
+            mock_resp_2 = Mock()
+            mock_resp_2.json.return_value = self.response_for_comment
+            with patch("requests.post", side_effect=[
```
Use single quotes here and below.

Also, I'm having a bit of trouble understanding this test. I suggest adding some comments explaining conceptually what you are trying to do.

SD-13 replied:
Replaced double quotes and added comments explaining the steps

seanlip commented on tests/index_test.py:
```diff
@@ -0,0 +1,218 @@
+                        '--max-wait-hours', '20',
+                        '--token', 'githubTokenForApiRequest'
+                    ])
+        self.assertTrue(request_1, self.response_for_discussions)
```
Not sure what you are asserting here -- assertTrue is generally only used for booleans, but this isn't one? Do you mean assertEqual?

SD-13 replied:
You are right. I wanted to use assertEqual, updated this time

seanlip commented on tests/index_test.py:
```diff
@@ -0,0 +1,218 @@
+
+
+class GenerateMessageTests(unittest.TestCase):
+    """test generate message function."""
```
Start docstrings with capital letters.

SD-13 replied:
Done here and elsewhere

seanlip commented on tests/index_test.py:
```diff
@@ -0,0 +1,218 @@
+
+    def test_generate_message(self):
+        
+        self.test_template = "{{ username }}\n{{ pr_list }}"
```
Use single quotes here and elsewhere, for strings.

SD-13 replied:
done

seanlip commented on tests/index_test.py:
```diff
@@ -0,0 +1,218 @@
+        with patch("builtins.open", file_data):
+            pr_list = '- [#123](https://githuburl.pull/123) [Waiting for the last 2 days, 8 hours]'
+            response = index.generate_message('reviewerName1', pr_list)
+        mocked_response = ('@reviewerName1\n'
```
This should be expected_response, right?

"Expected" is the thing that is used to check the result of your test against. "Mock" is data you supply to the test in order to keep the test bounded within the system it's testing.

SD-13 replied:
Thanks for explaining that, updated here and elsewhere.

seanlip reviewed (changes requested):
Thanks @SD-13. I took another pass, PTAL!

SD-13 committed:
some suggested changes and more

SD-13 committed:
more changes

SD-13 committed:
nit

SD-13 committed:
update docstring

SD-13 committed:
add comment for explanation

SD-13 commented:
@seanlip I have addressed your comments. Also, I need suggestions in a few places
1. I set the timeout value of a request to 15 seconds but I think we can reduce it, to maybe 5 seconds. what do you think?
2. Inside the `setUp` method in both the `main_test.py` and `github_services_test.py` we have pretty much the same code. Do you think we should keep it or get rid of it?

Please take another pass when you are free. Thanks!

SD-13 commented:
@seanlip FYI, after making those changes I have tested the functionality in my repos. The [`pending-review-notifier`](https://github.com/SD-13/pending-review-notifier) contains the code and  in [`test-pending-review-notifier`](https://github.com/SD-13/test-pending-review-notifier) repo, I used it as a GitHub action. To see the runs and the comments please check the `Actions` and the `Discussions` of [`test-pending-review-notifier`](https://github.com/SD-13/test-pending-review-notifier) repo. I changed the comment content to not bother reviewers. Thanks,

seanlip commented on src/github_domain.py:
```diff
@@ -20,41 +20,46 @@
             username: str. GitHub username of the assignee.
-            timestamp: datetime. The time when an event happened.
+            assigned_on_timestamp: datetime|None. The time when the reviewer was
+                assigned, else None.
```
...assigned, or None if [XYZ].

(What is XYZ? The reviewer was never assigned, or isn't assigned currently, or something else? Also it's not clear to me how this could possibly be None since then the assignee isn't an assignee. This needs to be explained more clearly, or the optional case should be removed.)

SD-13 replied:
Currently, the `Assignee` class is taking one required argument which is the `username`, and using `None` as the default value for `assigned_on_timestamp`.  The purpose of using `None` is to initialize it, nothing else. First, we are creating an `Assignee` instance in `from_github_response` and then using `update_assignee_timestamp` -> `__process_activity` -> `set_assigned_on_timestamp` to set the timestamp value. If there is an assignee only then an `Assignee` instance will be created with that username. What do you think about keeping `None` to initialize the argument with an explanation? I think making `assigned_on_timestamp` a default argument can be tricky because we need to make separate API calls to fetch specific PR timelines. FYI, Previously we were using `DEFAULT_TIMESTAMP: datetime.datetime = (
    datetime.datetime(datetime.MINYEAR, 1, 1, tzinfo=datetime.timezone.utc))` to initialize. I am ok with any of them but `None` is making more sense to me.

seanlip replied:
It would be better if we initialized it on creation. Can we get all the necessary info and only then create the Assignee object?

SD-13 replied:
Yeah, I think we can do that by modifying the functions in `github_services`. Signing off for the night, will try to do it tomorrow morning.

SD-13 replied:
I have updated and removed a few functions and methods and now we are creating the Assignee object at once.

seanlip commented on src/github_domain.py:
```diff
@@ -20,41 +20,46 @@
-        delta = datetime.datetime.now(datetime.timezone.utc) - self.timestamp
+
+        # Since a reviewer was assigned, we are not expecting the respective
+        # timestamp(when the the reviewer was assigned) to be none.
```
...timestamp (when ... (add space before '(')

the the --> the

SD-13 replied:
Just wanted to know, are you using any extension to catch this?

seanlip replied:
No, I'm just reading the PR "manually".

(I wanted to say "by hand" but I guess it's "by eye"? :P )

SD-13 replied:
done

seanlip commented on src/github_services.py:
```diff
@@ -76,17 +82,18 @@ def _get_request_headers() -> Dict[str, str]:
     org_name: str,
     repository: str,
-    wait_hours: int
+    max_waiting_time: int
```
The name of this variable needs to have a unit in it.

SD-13 replied:
updated

seanlip commented on src/github_services.py:
```diff
@@ -76,17 +82,18 @@ def _get_request_headers() -> Dict[str, str]:
-        wait_hours: int. The threshold to count a PR for discussion comment.
+        org_name: str. GitHub organization name.
+        repository: str. GitHub repository name.
+        max_waiting_time: int. The maximum time in hour to wait for a review. Any PR
```
hour --> hours

Any PR exceed ... --> If the waiting time for a PR review has exceeded this limit, the corresponding reviewer should be notified.

SD-13 replied:
done

seanlip commented on src/github_services.py:
```diff
@@ -121,12 +129,15 @@ def get_prs_assigned_to_reviewers(
                 continue
             for reviewer in pull_request.assignees:
+                # Since a reviewer was assigned, we are not expecting the respective
+                # timestamp(when the the reviewer was assigned) to be none.
```
Add space before '('.

the the --> the

SD-13 replied:
done

seanlip commented on src/github_services.py:
```diff
@@ -138,15 +149,21 @@ def __process_activity(
 ) -> None:
-    """Process activity and updates assignee timestamps.
-    """
+    """Process activity and update the respective timestamp when the assignee was assigned."""
```
What does "activity" refer to here? Please be more specific.

SD-13 replied:
I have changed the function name and the docstring. Please let me if that works.

seanlip commented on src/github_services.py:
```diff
@@ -221,13 +239,14 @@ def create_discussion_comment(
         json={'query': query, 'variables': variables},
-        headers=_get_request_headers()
+        headers=_get_request_headers(),
+        timeout=15
```
I suggest you make 15 into a named top-level constant and use that constant everywhere, so that it can be easily configured.

SD-13 replied:
done

seanlip commented on src/github_services.py:
```diff
@@ -251,10 +270,9 @@ def create_discussion_comment(
-    # know about some of their pending tasks. To learn more, check this out
+    # The following code is written in GraphQL and is being used to perform a mutation
+    # operation. More specifically, we are using it to comment in GitHub discussion to
+    # let reviewers know about some of their pending tasks. To learn more, check this out
```
check this out --> check out

(or, add a colon after "check this out")

SD-13 replied:
Added colon after "check this out".

seanlip commented on tests/main_test.py:
```diff
@@ -0,0 +1,281 @@
+            # Mock all get requests.
+            self.mock_all_get_requests(mock_request)
+
+            # Here we are mocking the two post requests, we will use in the below test.
```
...the two POST requests that we will use in the test below.

SD-13 replied:
done

seanlip commented on tests/main_test.py:
```diff
@@ -0,0 +1,281 @@
+            self.mock_all_get_requests(mock_request)
+
+            # Here we are mocking the two post requests, we will use in the below test.
+            # One request for fetching all existing GitHub Discussions data and the next
```
One request fetches all existing GitHub Discussions data, and the next request posts a comment in the ...

SD-13 replied:
done

seanlip commented on tests/main_test.py:
```diff
@@ -0,0 +1,281 @@
+            # request to post a comment in the particular GitHub Discussion.
+
+            # Creating a Mock instance.
+            mock_resp_1 = mock.Mock()
```
Can you change resp to response, here and below (if that's what it means)? Or to whatever the correct longer word is if it's not "response".

SD-13 replied:
done

seanlip commented on tests/main_test.py:
```diff
@@ -0,0 +1,281 @@
+            # One request for fetching all existing GitHub Discussions data and the next
+            # request to post a comment in the particular GitHub Discussion.
+
+            # Creating a Mock instance.
```
This and the next 2 comments aren't necessary. Comments should explain the "why" and "broader picture" behind the code, and not just say literally what the code does. Instead, the comment above should go above these three lines (and it's fine to drop the newline on line 243 so that it's clear the comment refers to this "batch").

SD-13 replied:
done.

seanlip commented on tests/main_test.py:
```diff
@@ -0,0 +1,281 @@
+            # Here we are patching the POST requests using side_effect. So, when you put
+            # callables inside `side_effect`, it will iterate through the items and
+            # return each at a time. For our test, we are expecting total 6 post requests,
+            # three for each(fetching discussions and posting comment) alternatively. To
```
Add space after "each".

SD-13 replied:
done

seanlip commented on tests/main_test.py:
```diff
@@ -0,0 +1,281 @@
+                        '--max-wait-hours', '20',
+                        '--token', 'githubTokenForApiRequest'
+                    ])
+        self.assertTrue(mock_resp_1.assert_called)
```
Missing pre-checks. See point 2 [here](https://github.com/oppia/oppia/wiki/Backend-tests#guidelines-for-writing-good-tests). Please make sure all other tests you write have pre-checks.

SD-13 replied:
Added

seanlip reviewed (changes requested):
The functionality looks great, thanks @SD-13!  Please add the proof of changes (i.e. your last comment) to the first comment of this PR thread so that it's easily accessible.

For the timeout value, I don't have a strong opinion. I think either is fine, I lean a bit towards 15 for more flexibility. The more important question is, what do we do if the timeout is exceeded -- does the Action fail and does anyone get notified?

For the setUp question, it's ok to leave it as you have it. (Though just to make sure I understand, what's your proposal exactly for the "get rid of it" alternative -- what are you getting rid of, and how will the test work if you do get rid of it?)

seanlip commented:
Thanks @SD-13, took a pass. Also added @DubeySandeep since he might want to look at this (as it is planned to replace what he originally built).

SD-13 committed:
sean's requested changes

SD-13 committed:
create Assignee at once

SD-13 committed:
add test in github_services

SD-13 commented:
@seanlip If the timeout value exceeds, it will fail noisily.
I didn't think about any alternative. Just wanted to know what you think before digging into it. Let's keep it then as it is.
Please have a look when you are free. Thanks,

seanlip commented on src/github_domain.py:
```diff
@@ -0,0 +1,133 @@
+            assignees: List(Assignee). List of assignees.
+        """
+        self.url = url
+        self.number = number
```
pr_number

(number is too generic, it can refer to anything)

SD-13 replied:
updated!

seanlip commented on src/github_services.py:
```diff
@@ -0,0 +1,299 @@
+PULL_REQUESTS_URL_TEMPLATE = 'https://api.github.com/repos/{0}/{1}/pulls'
+ISSUE_TIMELINE_URL_TEMPLATE = (
+    'https://api.github.com/repos/{0}/{1}/issues/{2}/timeline')
+TIMEOUT = 15
```
This should be TIMEOUT_SECS

SD-13 replied:
done

seanlip commented on tests/main_test.py:
```diff
@@ -0,0 +1,282 @@
+            mock_response_2 = mock.Mock()
+            mock_response_2.json.return_value = self.response_for_comment
+
+            self.assertTrue(mock_response_1.assert_not_called)
```
Probably these can go outside the "with" block so that they're on the same level as the post-checks?

SD-13 replied:
Moved outside the "with" block.

seanlip reviewed (changes requested):
Thanks @SD-13 -- just small comments left, but otherwise let's launch this! 

Just a question about the timeout exceeds -- you say it'll fail noisily, which is great. How does that "fail noisily" manifest? What actually happens -- does anyone get an email, is a discussion thread posted, etc.?

SD-13 committed:
few more changes

SD-13 commented:
Thanks @seanlip! For the timeout, if it fails, the error will look like [this](https://github.com/SD-13/test-pending-review-notifier/actions/runs/5694543688/job/15566708308). And just like we send test failure notifications in google chats using [send-webhook-notification](https://github.com/oppia/oppia/blob/dd4e4defab5abbba17bbc1d99f406d850ea0f68c/.github/workflows/pending-review-notification.yml#L30) action, we can do the same to notify folks. I am not fully sure if the `send-webhook-notification` action is working correctly or if it send notification in chats, please correct me if I am wrong but I think I heard this in a meeting.

seanlip reviewed (approved):
OK, sg, thanks @SD-13. I think if you hook it up using whatever the existing notification mechanism is, it should probably be fine.

Feel free to launch this and then if it looks like it's working correctly after a couple weeks then we can take down the old one.

Thanks!

seanlip commented:
Also @SD-13 I suggest that you coordinate with the dev workflow leads about how to manage the notifications that will result when this is turned on -- maybe draw up a launch plan with steps. Otherwise GitHub Discussions is going to get quite noisy with unresolved conversations. So there probably should be some kind of process (ideally documented) around this.

/cc @U8NWXD @DubeySandeep

SD-13 commented:
@U8NWXD @shivkant5 @DubeySandeep Initially I am thinking about locking the discussion so that other folks can't comment there. To make it less noisy it would be better if we hide the existing comments before posting new ones but I am not getting a way to do it using the API. I think we have two options, one is to keep the previous comments just like the team discussion and the other is to delete earlier comments before posting new ones. Since this discussion is public, I am more toward the latter (deleting existing ones before). Please let me know what you think about this. Thanks!

U8NWXD commented:
Hmm I agree on locking. This thread isn't going to be monitored, so we don't want people asking for help on it.

Regarding deleting comments, I think it could be useful to keep a little bit of history (maybe a month or two) so we can spot trends in review times, but beyond that I agree with deleting

SD-13 commented:
@U8NWXD sounds good. I will start working to implement it then. Thanks!

SD-13 committed:
add delete comment functionality

SD-13 commented:
@seanlip @U8NWXD I have added the changes to delete comments and tested in [this GitHub action run](https://github.com/SD-13/test-pending-review-notifier/actions/runs/5694543688/job/15704593480). Please have a look. In the meantime, I will run the linter from `oppia/oppia` repo.

SD-13 committed:
fix lint checks

SD-13 committed:
nits

SD-13 commented on .github/PENDING_REVIEW_NOTIFICATION_TEMPLATE.md:
```diff
@@ -0,0 +1,8 @@
+It looks like you're assigned to this PR, but haven't taken any action for at least 2 days:
+{{ pr_list }}
+
+Please review and unassign yourself from the pending PRs as soon as possible, and reply to this thread once all the PRs are reviewed.
```
```suggestion
Please review and unassign yourself from the pending PRs as soon as possible.
```

SD-13 replied:
@U8NWXD Do you agree that we should make this change since we are not expecting reviewers to reply?

U8NWXD replied:
Yep, I agree! Let me know when you want me to merge this PR

SD-13 replied:
Let's merge this today!

SD-13 committed:
update template

SD-13 commented:
@U8NWXD @seanlip I saw this action failed in the `oppia/oppia` and after investigating I found that, it was a mistake in the query being used to fetch the existing discussions and it didn't get caught in my test repo because of less no of discussions( I had only two discussions). I will open a PR to fix it. Thanks!
